### PR TITLE
EVG-13441 Validate project modules set branch field

### PR DESCRIPTION
**Pull request by rayangler**
[[Jira Ticket](https://jira.mongodb.org/browse/EVG-13441)]

### CODE REVIEW

**Review by malikchaya2 — APPROVED** [review]

File: validator/project_validator.go [hunk 1]
```diff
@@ -584,6 +585,20 @@ func validateTaskNames(project *model.Project) ValidationErrors {
 	return errs
 }
 
+// validateModuleBranches checks to make sure that branches of modules have a set branch.
+func validateModuleBranches(project *model.Project) ValidationErrors {
+	errs := ValidationErrors{}
+	for _, module := range project.Modules {
+		if module.Branch == "" {
```
Comment (ybrill) [inline validator/project_validator.go:592]: maybe once we're validating modules we should also validate the name and repo fields?

Comment (rayangler) [reply validator/project_validator.go:592]: Sounds good to me. Would we only be checking that the name and repo also aren't empty strings? Or are there other requirements for a valid name and repo?

File: validator/project_validator.go [hunk 2]
```diff
@@ -584,6 +585,20 @@ func validateTaskNames(project *model.Project) ValidationErrors {
...
+			errs = append(errs, ValidationError{
+				Level:   Warning,
+				Message: fmt.Sprintf("module '%s' should have a set branch", module.Name),
+			})
+		}
+	}
+	return errs
```
Comment (malikchaya2) [inline validator/project_validator.go:599]: Oh, one question: I realize that this is an error not a warning. @ybrill can this affect evergreen yamls in use?

Comment (ybrill) [reply validator/project_validator.go:599]: I thought it _is_ just a warning, and that this is the function of the Level field.

Comment (malikchaya2) [reply validator/project_validator.go:599]: ah yes, you are right. thanks!

File: validator/project_validator.go (hunk 1 above)
Comment (ybrill) [reply validator/project_validator.go:592]: For name it makes sense to me to validate 1) it's not the empty string (the empty string is a reserved module name), as you suggest. 2) there are no duplicates.
For repo at least that 1) it's not the empty string. Maybe also 2) that it conforms to one of the [Git URLs](https://git-scm.com/docs/git-clone#_git_urls), though the effort required may not be worth the benefit.

Comment (ybrill) [reply validator/project_validator.go:592]: It looks like we already have [a function for git URL validation](https://github.com/evergreen-ci/evergreen/blob/08d8cc896f586d8a23edf9bab940d6f2d3e70ffa/thirdparty/git.go#L268)

File: validator/project_validator.go [hunk 3]
```diff
@@ -584,6 +586,48 @@ func validateTaskNames(project *model.Project) ValidationErrors {
 	return errs
 }
 
+// validateModules checks to make sure that the module's name, branch, and repo are correct.
+func validateModules(project *model.Project) ValidationErrors {
+	errs := ValidationErrors{}
+	moduleNames := map[string]bool{}
+
+	for _, module := range project.Modules {
+		// Warn if name is a duplicate
```
Comment (ybrill) [inline validator/project_validator.go]: what about when the name is empty?

File: validator/project_validator_test.go [hunk 4]
```diff
@@ -700,6 +700,100 @@ func TestValidateTaskNames(t *testing.T) {
 	})
 }
 
+func TestValidateModules(t *testing.T) {
+	Convey("When validating a project's modules", t, func() {
+		Convey("An error should be returned when more than one module shares the same name", func() {
```
Comment (ybrill) [inline validator/project_validator_test.go:705]: or if it's empty

**Review by ybrill — APPROVED** [review]

**Pull request merged**